tools/c7n_azure/tests/azure_common.py

```diff
@@ -105,22 +130,103 @@ def request_callback(self, request):
+        if not content_type or 'application/json' not in content_type:
+            return response
+
+        body = response['body'].pop('string').decode('utf-8')
+
+        # Clean up subscription IDs in body
+        body = AzureVCRBaseTest._replace_subscription_id(body)
+
+        response['body']['data'] = json.loads(body)
+
+        # Replace Graph API responses
+        odata_metadata = response['body']['data'].get('odata.metadata', None)
+        if odata_metadata and "directoryObjects" in odata_metadata:
+            response['body']['data'] = GRAPH_RESPONSE
+
+        return response
+
+    @staticmethod
+    def _replace_subscription_id(s):
+        if "subscriptions" in s:
+            return re.sub(
+                r"(?P<prefix>(/|%2F)subscriptions(/|%2F))"
+                r"[\da-zA-Z]{8}-([\da-zA-Z]{4}-){3}[\da-zA-Z]{12}",
+                r"\g<prefix>" + DEFAULT_SUBSCRIPTION_ID, s)
+        return s
+
+    @staticmethod
+    def _replace_tenant_id(s):
+        if "graph.windows.net" in s:
+            return re.sub(
+                r"(?P<prefix>(/|%2F)graph.windows.net(/|%2F))"
+                r"[\da-zA-Z]{8}-([\da-zA-Z]{4}-){3}[\da-zA-Z]{12}",
+                r"\g<prefix>" + DEFAULT_TENANT_ID, s)
+        return s
+
 
 class BaseTest(TestUtils, AzureVCRBaseTest):
     """ Azure base testing class.
     """
+    def __init__(self, *args, **kwargs):
+        super(BaseTest, self).__init__(*args, **kwargs)
+        self._requires_polling = False
 
     def setUp(self):
         super(BaseTest, self).setUp()
         ThreadHelper.disable_multi_threading = True
 
-        # Patch Poller with constructor that always disables polling
-        self.lro_patch = patch.object(msrest.polling.LROPoller, '__init__', BaseTest.lro_test_init)
-        self.lro_patch.start()
+        if not self._requires_polling:
+            # Patch Poller with constructor that always disables polling
+            # This breaks blocking on long running operations (resource creation).
+            self._lro_patch = patch.object(msrest.polling.LROPoller, '__init__', BaseTest.lro_init)
+            self._lro_patch.start()
+            self.addCleanup(self._lro_patch.stop)
+        elif hasattr(self, 'cassette') or os.path.isfile(self.cassette._path):
+            # If using polling we need to monkey patch the timeout during playback
+            # or we'll have long sleeps introduced into our test runs
+            Session._old_client = Session.client
+            Session.client = BaseTest.session_client_wrapper
+            self.addCleanup(BaseTest.session_client_cleanup)
+
+        self._utc_patch = patch.object(actions, 'utcnow', BaseTest.get_test_date)
+        self._utc_patch.start()
+        self.addCleanup(self._utc_patch.stop)
 
-    def tearDown(self):
-        super(BaseTest, self).tearDown()
-        self.lro_patch.stop()
+        if constants.ENV_ACCESS_TOKEN in os.environ:
```

Comment:
not sure if it is relevant, but I'd prefer to make this patching when we replay a cassette.

Reply:
agreed, fixed.